persistence/src/main/scala/akka/persistence/EventStoreinterop.scala

```diff
@@ -0,0 +1,196 @@
+    *   internal event buffer capacity, on oveflow will raise [[BufferOverflowException]]
+    * @param journalPluginId
+    *   Akka persistence journal plugin ID
+    * @param eventSourcedId
+    *   Akka persistence ID, unique per each actor
+    * @return
+    *   instance of [[EventStore]]
+    */
+  def apply[F[_]: Async: FromFuture: ToTry, A](
+    system: ActorSystem,
+    timeout: FiniteDuration,
+    capacity: Int,
+    journalPluginId: String,
+    eventSourcedId: EventSourcedId
+  ): F[EventStore[F, A]] =
+    Sync[F]
+      .delay {
+        val actorRef = Persistence(system).journalFor(journalPluginId)
+        ActorEffect.fromActor(actorRef)
+      }
+      .map { journaller =>
+        new EventStore[F, A] {
+
+          import sstream.FoldWhile._
+
+          val persistenceId = eventSourcedId.value
+
+          override def events(fromSeqNr: SeqNr): F[sstream.Stream[F, EventStore.Persisted[A]]] = {
+
+            type Buffer = Vector[EventStore.Event[A]]
+
+            def actor(buffer: Ref[F, Buffer]) =
+              LocalActorRef[F, Unit, SeqNr]({}, timeout) {
+
+                case (_, JournalProtocol.ReplayedMessage(persisted)) =>
+                  if (persisted.deleted) ().asLeft[SeqNr].pure[F]
+                  else {
+                    val payload = persisted.payload.asInstanceOf[A]
+                    val event   = EventStore.Event(payload, persisted.sequenceNr)
+
+                    buffer
+                      .modify { buffer =>
+                        val buffer1 = buffer :+ event
+                        val lenght  = buffer.length + 1
+                        buffer1 -> lenght
+                      }
+                      .flatMap { lenght =>
+                        if (lenght > capacity) {
+                          new BufferOverflowException(capacity, persistenceId).raiseError[F, Either[Unit, SeqNr]]
+                        } else {
+                          ().asLeft[SeqNr].pure[F]
+                        }
+                      }
+                  }
+
+                case (_, JournalProtocol.RecoverySuccess(seqNr)) =>
+                  seqNr.asRight[Unit].pure[F]
+
+                case (_, JournalProtocol.ReplayMessagesFailure(error)) =>
+                  error.raiseError[F, Either[Unit, SeqNr]]
+              }
+
+            for {
+              buffer <- Ref[F].of[Buffer](Vector.empty)
+              actor  <- actor(buffer)
+              request = JournalProtocol.ReplayMessages(fromSeqNr, SeqNr.Max, Long.MaxValue, persistenceId, actor.ref)
+              _      <- journaller.tell(request)
+            } yield new sstream.Stream[F, EventStore.Persisted[A]] {
+
+              override def foldWhileM[L, R](l: L)(f: (L, EventStore.Persisted[A]) => F[Either[L, R]]): F[Either[L, R]] =
+                l.asLeft[R]
+                  .tailRecM {
+
+                    case Left(l) =>
+                      for {
+                        events <- buffer.getAndSet(Vector.empty)
```

Comment:
I feel a bit of uneasy about having to access `buffer` directly, like this. It breaks encapsulation of the logic found in `#actor(buffer: Ref[F, Buffer])` method a bit.

I wonder if it could be beneficial to extract `#actor(buffer: Ref[F, Buffer])` to a separate class, and then have the method on it, which would call this `buffer.getAndSet` method and do not allow direct access to buffer from outside.

Feel free to ignore if you think it is a bad idea.

Reply:
https://github.com/evolution-gaming/akka-effect/pull/276#discussion_r1450691747